modules/generator/processor/localblocks/processor.go

```diff
@@ -387,6 +391,78 @@ func (p *Processor) GetMetrics(ctx context.Context, req *tempopb.SpanMetricsRequ
 	return resp, nil
 }
 
+// QueryRange returns metrics.
+func (p *Processor) QueryRange(ctx context.Context, req *tempopb.QueryRangeRequest) (traceql.SeriesSet, error) {
+	p.blocksMtx.RLock()
+	defer p.blocksMtx.RUnlock()
+
+	cutoff := time.Now().Add(-p.Cfg.CompleteBlockTimeout).Add(-timeBuffer)
+	if req.Start < uint64(cutoff.UnixNano()) {
+		return nil, fmt.Errorf("time range must be within last %v", p.Cfg.CompleteBlockTimeout)
+	}
+
+	// Blocks to check
+	blocks := make([]common.BackendBlock, 0, 1+len(p.walBlocks)+len(p.completeBlocks))
+	if p.headBlock != nil {
+		blocks = append(blocks, p.headBlock)
+	}
+	for _, b := range p.walBlocks {
+		blocks = append(blocks, b)
+	}
+	for _, b := range p.completeBlocks {
+		blocks = append(blocks, b)
+	}
+
+	eval, err := traceql.NewEngine().CompileMetricsQueryRange(req, false)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		wg     = boundedwaitgroup.New(10)
+		jobErr = atomic.Error{}
+	)
+
+	for _, b := range blocks {
+
+		start := uint64(b.BlockMeta().StartTime.UnixNano())
+		end := uint64(b.BlockMeta().EndTime.UnixNano())
+		if start > req.End || end < req.Start {
+			// Out of time range
+			continue
+		}
+
+		wg.Add(1)
+		go func(b common.BackendBlock) {
+			defer wg.Done()
+
+			span, ctx2 := opentracing.StartSpanFromContext(ctx, "Processor.QueryRange.Block", opentracing.Tags{
+				"block":     b.BlockMeta().BlockID,
+				"blockSize": b.BlockMeta().Size,
+			})
+			defer span.Finish()
+
+			// TODO - caching
+			f := traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+				return b.Fetch(ctx, req, common.DefaultSearchOptions())
+			})
+
+			err := eval.Do(ctx2, f)
+			if err != nil {
+				jobErr.Store(err)
```

Comment:
Alternatively we could append to a slice to get all of the errors.  I think this just stores the last error.

Reply:
I think returning just the last error is ok.  Now looking at it more closely, `QueryRange` ideally exits immediately after the first error and cancels all remaining work.